pkg/controller/pxc/config.go

```diff
@@ -0,0 +1,301 @@
+		return controllerutil.OperationResultNone, errors.Wrap(err, "delete config map")
+	}
+
+	configMap := config.NewConfigMap(cr, haproxyConfigName, "haproxy-global.cfg", cr.Spec.HAProxy.Configuration)
+
+	err := k8s.SetControllerReference(cr, configMap, r.scheme)
+	if err != nil {
+		return controllerutil.OperationResultNone, errors.Wrap(err, "set controller ref")
+	}
+
+	res, err := createOrUpdateConfigmap(ctx, r.client, configMap)
+	if err != nil {
+		return res, errors.Wrap(err, "create or update config map")
+	}
+
+	return res, nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) reconcileLogcollectorConfigMap(ctx context.Context, cr *api.PerconaXtraDBCluster) (controllerutil.OperationResult, error) {
+	logCollectorConfigName := config.CustomConfigMapName(cr.Name, "logcollector")
+
+	if cr.Spec.LogCollector == nil || cr.Spec.LogCollector.Configuration == "" {
+		err := deleteConfigMapIfExists(ctx, r.client, cr, logCollectorConfigName)
+		return controllerutil.OperationResultNone, errors.Wrap(err, "delete config map")
+	}
+
+	configMap := config.NewConfigMap(cr, logCollectorConfigName, "fluentbit_custom.conf", cr.Spec.LogCollector.Configuration)
+
+	err := k8s.SetControllerReference(cr, configMap, r.scheme)
+	if err != nil {
+		return controllerutil.OperationResultNone, errors.Wrap(err, "set controller ref")
+	}
+
+	res, err := createOrUpdateConfigmap(ctx, r.client, configMap)
+	if err != nil {
+		return res, errors.Wrap(err, "create or update config map")
+	}
+
+	return res, nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) createHookScriptConfigMap(ctx context.Context, cr *api.PerconaXtraDBCluster, hookScript string, configMapName string) error {
+	configMap := config.NewConfigMap(cr, configMapName, "hook.sh", hookScript)
+
+	err := k8s.SetControllerReference(cr, configMap, r.scheme)
+	if err != nil {
+		return errors.Wrap(err, "set controller ref")
+	}
+
+	_, err = createOrUpdateConfigmap(ctx, r.client, configMap)
+	if err != nil {
+		return errors.Wrap(err, "create or update configmap")
+	}
+
+	return nil
+}
+
+func createOrUpdateConfigmap(ctx context.Context, cl client.Client, configMap *corev1.ConfigMap) (controllerutil.OperationResult, error) {
+	log := logf.FromContext(ctx)
+
+	currMap := &corev1.ConfigMap{}
+	err := cl.Get(ctx, types.NamespacedName{Namespace: configMap.Namespace, Name: configMap.Name}, currMap)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return controllerutil.OperationResultNone, errors.Wrap(err, "get current configmap")
+	}
+
+	if k8serrors.IsNotFound(err) {
+		log.V(1).Info("Creating object", "object", configMap.Name, "kind", configMap.GetObjectKind())
+		return controllerutil.OperationResultCreated, cl.Create(ctx, configMap)
+	}
+
+	if !reflect.DeepEqual(currMap.Data, configMap.Data) {
+		log.V(1).Info("Updating object", "object", configMap.Name, "kind", configMap.GetObjectKind())
+		return controllerutil.OperationResultUpdated, cl.Update(ctx, configMap)
```

Comment:
Can we have conflicts when updating the configmap needing to retry?

Reply:
unlikely but not impossible